tests/unittests/OnnxImporterTest.cpp

```diff
@@ -27,6 +27,174 @@
+      const std::string &inputName = pvalueInfo->name();
+      if (inputName != tensorNames[i]) {
+        continue;
+      }
+      std::string new_name = "dummy_input" + std::to_string(i);
+      pvalueInfo->set_name(new_name);
+      auto RH = tensors[i]->getHandle<>();
+      ONNX_NAMESPACE::TensorProto *tp = gp->add_initializer();
+      tp->set_name(tensorNames[i]);
+      for (int k = 0; k < tensors[i]->dims().size(); k++) {
+        tp->add_dims(tensors[i]->dims()[k]);
+      }
+      switch (RH.getElementType()) {
+      case ElemKind::FloatTy:
+        tp->set_data_type(ONNX_NAMESPACE::TensorProto::FLOAT);
+        for (int k = 0; k < tensors[i]->size(); k++) {
+          tp->add_float_data(RH.raw(k));
+        }
+        break;
+      case ElemKind::Int64ITy:
+        tp->set_data_type(ONNX_NAMESPACE::TensorProto::INT64);
+        for (int k = 0; k < tensors[i]->size(); k++) {
+          tp->add_int64_data(RH.raw(k));
+        }
+        break;
+      case ElemKind::Int32ITy:
+        tp->set_data_type(ONNX_NAMESPACE::TensorProto::INT32);
+        for (int k = 0; k < tensors[i]->size(); k++) {
+          tp->add_int32_data(RH.raw(k));
+        }
+        break;
+      default:
+        std::cout << "Unsupported datatype";
+        return false;
+      }
+    }
+  }
+  gp->clear_input();
+  return true;
+}
+
+bool checkConstFoldedOutput(std::string NetFilename,
+                            llvm::ArrayRef<const char *> tensorNames,
+                            llvm::ArrayRef<Tensor *> tensors,
+                            llvm::ArrayRef<Tensor *> ref_tensors) {
+  ONNX_NAMESPACE::ModelProto modelDef;
+  std::string CFNetFilename = "dummy.onnxtxt";
+  auto lhsOrErrV = loadProto(NetFilename);
+  static_assert(IsLLVMExpected<decltype(lhsOrErrV)>(),
+                "Expected value to be a llvm::Expected");
+  if (!lhsOrErrV) {
+    return false;
+  }
+  modelDef = std::move(lhsOrErrV.get());
+  if (!replacePlaceholderWithConstant(modelDef, tensorNames, tensors))
+    return false;
+  if (!saveProto(CFNetFilename, modelDef))
+    return false;
+  setConstantFoldLoaderOpsFlag(true);
+  PlaceholderBindings binding_temp;
+  ExecutionEngine ee_temp;
+  Module &mod_temp = ee_temp.getModule();
+  Function *pfunc_temp = mod_temp.createFunction("temp");
+  ONNXModelLoader onnxLD(CFNetFilename, {}, {}, *pfunc_temp);
+  setConstantFoldLoaderOpsFlag(false);
+
+  const llvm::StringMap<Placeholder *> pOut = onnxLD.getOutputVarsMapping();
+  for (int i = 0; i < modelDef.graph().output_size(); i++) {
+    NodeValue nv;
+    auto lhsOrErrV =
+        onnxLD.getNodeValueByName(modelDef.graph().output(i).name());
+    static_assert(IsLLVMExpected<decltype(lhsOrErrV)>(),
+                  "Expected value to be a llvm::Expected");
+    if (lhsOrErrV) {
+      nv = std::move(lhsOrErrV.get());
+      switch (nv.getElementType()) {
+      case ElemKind::FloatTy:
+        EXPECT_TRUE(ref_tensors[i]->getHandle().dims() ==
```

Comment:
Tiny nit: Each of the case blocks does the same, just the type differs. Looks like a lot of boilerplate code. May be a you could use a template helper and call it with different types?

Reply:
I just discovered isEqual member function of Tensor, so this whole code will go.

Reply:
done

> Looks pretty good -- some nits, and one main question about the necessity of `FunctionState`.
> 
> Also, can you also add a unit test for testing an op like TopK with multiple outputs?

@jfix71 I have a separate patch for TopK, which I am planning to submit as a separate MR to add support for k as an input instead of just attribute. With that folding test for TopK will be available.